ganttproject/src/main/java/net/sourceforge/ganttproject/GanttProject.java

```diff
@@ -956,33 +955,73 @@ public void refresh() {
     super.repaint();
   }
 
-  // TODO: Accumulate changes instead of sending it every time.
+  private interface TxnSendListener {
+    void onSendCompleted();
+  }
+
+  private final AtomicReference<TxnSendListener> txnSendingListener = new AtomicReference<>();
+
   private Unit sendProjectStateLogs() {
     gpLogger.debug("Sending project state logs");
+    if (txnSendingListener.get() != null) return Unit.INSTANCE;
+    var baseTxnCommitInfo = myBaseTxnCommitInfo.get();
+    if (baseTxnCommitInfo.left == null) {
+      // Connection with the server was not established.
+      return Unit.INSTANCE;
+    }
+    List<XlogRecord> transactions;
     try {
-      var baseTxnCommitInfo = myBaseTxnCommitInfo.get();
-      var txns = myProjectDatabase.fetchTransactions(baseTxnCommitInfo.right + 1, 1);
-      if (!txns.isEmpty()) {
+      transactions = myProjectDatabase.fetchTransactions(baseTxnCommitInfo.right + 1, 1);
+    } catch (ProjectDatabaseException e) {
+      gpLogger.error("Failed to send logs", new Object[]{}, ImmutableMap.of(), e);
+      return Unit.INSTANCE;
+    }
+    if (!transactions.isEmpty()) {
+      var listener = new TxnSendListener() {
+        @Override
+        public void onSendCompleted() {
+          txnSendingListener.compareAndSet(this, null);
+        }
+      };
+      if (txnSendingListener.compareAndSet(null, listener)) {
         getWebSocket().sendLogs(new InputXlog(
           baseTxnCommitInfo.left,
           "userId",
           "refid",
-          txns
+          transactions
         ));
+      } else {
+        // Logs were sent by another thread, no action required.
       }
-    } catch (ProjectDatabaseException e) {
-      gpLogger.error("Failed to send logs", new Object[]{}, ImmutableMap.of(), e);
+    }
+    return Unit.INSTANCE;
+  }
+
+  @Override
+  protected Unit onProjectLogUpdate() {
+    if (isColloboqueLocalTest()) {
+      super.onProjectLogUpdate();
+      sendProjectStateLogs();
     }
     return Unit.INSTANCE;
   }
 
   private Unit fireXlogReceived(ServerCommitResponse response) {
     myBaseTxnCommitInfo.update(response.getBaseTxnId(), response.getNewBaseTxnId(), 1);
+    txnSendingListener.get().onSendCompleted();
+    sendProjectStateLogs();
     return Unit.INSTANCE;
   }
 
+  // TODO: sync logs with the server.
   private Unit onBaseTxnIdReceived(String baseTxnId) {
-    myBaseTxnCommitInfo.update("", baseTxnId, 0);
+    // Websocket is [re-]started. Previous messages are discarded.
+    txnSendingListener.set(null);
+    if (EMPTY_LOG_BASE_TXN_ID.equals(baseTxnId)) {
+      myBaseTxnCommitInfo.reset();
+      myBaseTxnCommitInfo.update(null, baseTxnId, 0);
+      sendProjectStateLogs();
```

Comment:
remove it? Sending logs doesn't make too much sense if we just have reset the state.

Reply:
In such a case, we'll send all the logs from the beginning, one by one
Where should we start sending logs instead? I thought that it's the right place

Reply:
Send the logs when an undoable edit happens. But I don't think that a user has any chance to make an undoable edit immediately at the moment when we reset the log.